### DISCUSSION TITLE
Chariots(車) was used very late in the game for Xiangqi

### DISCUSSION THREAD

**Discussion by kappa-wingman** [fairy-stockfish/Fairy-Stockfish#233] [Analysis]
Greetings,

I saw this discussion https://github.com/ianfab/Fairy-Stockfish/issues/230#issuecomment-753011482
Yes, Fairy-Stockfish use chariot very late. Below is a game with another engine called Challenger. Despite that engine use chariots early, it lose. Fairy-SF used the chariots at step 14 or 15 but still wins (Opponent perpetual chasing)! 

```
[Event "Computer Chess Game"]
[Site "DESKTOP"]
[Date "2021.01.02"]
[Round "-"]
[White "Challenger 010121 (UCI2WB)"]
[Black "Fairy-Stockfish 271220 LB"]
[Result "0-1"]
[TimeControl "40/300"]
[Variant "xiangqi"]
[Annotator "1. +0.49   1... +0.09"]

1. Hc2 g5 2. Ra1 Ege7 3. Ce2 Hc7 4. Hg2 Afe8 5. Rh0 Cf7 6. Rf1 Hg7 7. c4
Ca7 8. Hb4 c5 9. cxc5 Exc5 10. Rh6 Cxa3 11. e4 Ce7 12. Rf3 Cxe4+ 13. Afe1
Ca4 14. Rh7 Rg9 15. Cc2 Rb9 16. Ha2 Hd5 17. Rd3 Cec4 18. Cxc5 Hf4 19. Rf3
a5 20. Cf2 Rb5 21. Cc6 Rb6 22. Cxi6 e5 23. Ege2 Cd4 24. Cxf4 Rxi6 25. Cxa4
axa4 26. Rh4 Cd7 27. Hc3 a3 28. Hb5 Re6 29. Rc4 Hi6 30. i4 Ri9 31. i5 Hh8
32. Rxc9 Rxi5 33. He3 e4 34. Hc2 ab3 35. Ha7 Ri0+ 36. Af0 b2 37. Hb4 e3 38.
Rf2 Hf7 39. Hbc6 bc2 40. Hd4 exe2 41. Exe2 Re4 42. Hxc2 Cd2 43. Rf1 Rxe2+
44. Ae1 Re4 45. Rf2 Ad7 46. Rf1 Cg2 47. Rg1 Rf4 48. Kd0 Ri2 49. Rc7 A7e8
50. Hc8+ Kf9 51. He7 Ch2 52. Rc9 Kf8 53. Hd5 Rf3 54. Hce3 Ch5 55. Ke0 Ra2
56. Rc0 Ra5 57. Rd0 Ch6 58. g4 Ce6 59. Hc7 Ce7 60. Hd5 Ce5 61. Rd2 Ra0+ 62.
Rd0 Hh6 63. Rg0 Ra5 64. He7 Ra6 65. Rd5 Ce4 66. Ri0 Hxg4 67. Ra5 Raf6 68.
Kd0 Hxe3 69. Rd5 Rb6 70. Hg6+ Rxg6 71. Rd3 Rc6 72. Rxe3 Rd6+ 73. Ke0 Rxe3
74. Ri8+ Kf9 75. Ri9+ Kf8 76. Ri8+ Kf9 77. Ri9+ Kf8 78. Ri8+
{Black wins} 0-1
```

**Comment by SGQuek** [comment]
To @kappa-wingman 
Thank you very much for the confirmation. Lets hope that when the future versions of SF becomes available on Github, it will be able to defeat Challenger by checkmate(將死) or stalemate(困斃).

**Comment by SGQuek** [comment]
Moreover, based on my knowledge and experience dealing with neural network based AI model. The timing of deploying a piece, such as chariots, should be very visible to any neural network based design. So an NNUE will definitely be able to recognize such distinctive pattern and make the appropriate judgement. 

Hopefully a version utilizing NNUE will be available to all of us eventually, just like the original Stockfish 12.

**Comment by ianfab** [comment]
I just noticed that the definition of (semi-)open files in Fairy-Stockfish so far only considers pawns and shogi pawns, not soldiers, which means that rooks do not get an extra bonus in Xiangqi when they are on an open file. I will test if changing this yields a playing strength improvement for Xiangqi (and Janggi). If it passes, this might help to mitigate this issue.

**Reply by ianfab** [reply]
https://github.com/ianfab/Fairy-Stockfish/commit/92ea16e4dcc49815dc167a069db9c14d8c009e00 is merged to master now. I hope this will help, at least it seems to improve playing strength, and speaking from the code it should encourage Fairy-SF to put rooks on open files.

**Reply by kappa-wingman** [reply]
I just tried the new version against two different engines. The new versions uses Chariots/rook early.

1) GGZero (very strong, free binary in here: https://github.com/leedavid/leela-chess-to-Chinese-Chess/releases/tag/v0.72 )

```
[Event "Computer Chess Game"]
[Site "Desktop"]
[Date "2021.01.06"]
[Round "-"]
[White "GGchess 05042018_MAXCPU:36 (UCI2WB)"]
[Black "Fairy-Stockfish 060121 LB"]
[Result "1-0"]
[TimeControl "40/300"]
[Variant "xiangqi"]
[Annotator "1. +0.37   1... +0.44"]

1. Che2 Hc7 2. Hg2 g5 3. Hc2 Ade8 4. Ca2 Rb9 5. Rb0 Cb3 6. Rh0 Cg7 7. Rh8
Cxg3 8. Hh0 Cb8 9. Rh3 Cxc3 10. e4 Cc5 11. Hd4 Cb2 12. Hxe6 Hxe6 13. Cxe6+
Ece7 14. Rh2 Hg7 15. Cexa6 Cb1 16. Rh1 Cb6 17. Rhb1 Rh9 18. Hg2 Rh3 19.
Ca9+ Rxa9 20. Cxa9 Rxa3 21. Rxb6 Rxa9 22. e5 Ra4 23. R0b4 Rxb4 24. Rxb4 Ca5
25. Rb9+ Ad9 26. Rb6 c5 27. Rg6 He8 28. e6 i5 29. exe7 Exe7 30. Re6 Eg9 31.
Hh4 Ca7 32. Hf5 Ce7+ 33. Hxe7 Exe7 34. Rxe7 g4 35. Rc7 g3 36. Rxc5 gf3 37.
Rc3 f2 38. Rc2 f1 39. Rg2 fxf0+ 40. Ke1 fe0 41. Kxe0 i4
{White wins} 1-0
```

2) Challenger

```
[Event "Computer Chess Game"]
[Site "DESKTOP"]
[Date "2021.01.06"]
[Round "-"]
[White "Challenger 010121 (UCI2WB)"]
[Black "Fairy-Stockfish 060121 LB"]
[Result "0-1"]
[TimeControl "40/300"]
[Variant "xiangqi"]
[Annotator "1. +0.49   1... -0.05"]

1. Hc2 Hg7 2. g4 Hc7 3. Hg2 c5 4. Ri1 Ch3 5. Hf4 Cxc3 6. He2 Cd3 7. Rd1 Cd5
8. Cb6 Afe8 9. Cc6 Cd7 10. Rb0 Rb9 11. Rb4 Ege7 12. Rf1 Rh9 13. Cg2 Rh5 14.
Cgxg6 Rd5 15. Rff4 e5 16. Rf6 Ca7 17. Rbf4 a5 18. Hb4 c4 19. Rxc4 Cxa3 20.
Cb6 a4 21. Hc2 Ha8 22. Rxa4 Hxb6 23. Rxa3 Hc4 24. Rc3 Rd4 25. Ea2 Ha5 26.
Ra3 Rd5 27. Cg5 Rd2 28. Cxa5 Rb2 29. g5 Rdxc2 30. gf5 Cd1 31. Rg6 Hf9 32.
fxe5 Cb1 33. Rc3 Rxc3 34. Hxc3 Cb0+ 35. Ec0 Rc2 36. He2 Rxc0 37. Rb6 Ca0
38. Cd5 Hh8 39. e4 Rc4+ 40. Ae1 Rxe4 41. Ra6 Cb0 42. Rb6 Ca0 43. Ra6 Cb0
44. Rb6 Cc0 45. Cd8 Hf7 46. Rc6 Cc4 47. Rxc9 Hxe5 48. Rc7 Re3 49. Rxe7 Ce4
50. Rb7 Rc3 51. Rb4 Rc0+ 52. Cd0 Hc4 53. Rb1 Ce7 54. Rd1 Rc2 55. Ra1 Hd2+
56. Rd1 He4 57. Hf4 Hf6+ 58. Ad2 Hg4 59. Hg6 Ce4 60. Hf4 Hf2+ 61. Rf1 Rxd2
62. Cb0 Af7 63. Cd0 Rb2 64. Cc0 Ade8 65. Cd0 Ce3 66. Hd5 Ce5 67. Cd1 He4+
68. He3 Rb0+ 69. Cd0 Rb3 70. Ae1 Rxe3 71. Cd2 Rb3 72. Rf5 Ce6 73. Kf0 Rb0+
74. Cd0 Cxe1 75. Rd5 Ci1 76. Ei2 Ca1 77. Rd1 Ca0 78. Kf1 Cxd0 79. Eg0 Rb3
80. Rxd0 Re3 81. Kf0 Ad9 82. Rd2 Hg3 83. Re2 Hxe2+ 84. Exe2 Rxe2 85. Kf1
Re4 86. i4 Rf4#
{White resigns} 0-1
```

But personally, I think Fairy SF engine favors the use of Pao/Cannon to attack early and/or moved too much (including the deploy of Advisor before the remaining Horse. Questionable, maybe it's necessary to deploy the Advisor at that time). That would delayed the deploy of the remaining Horse and Chariots/Rook.

It's about utilizing all the pieces for the opening as soon as possible (for greater overall efficiency), which I learned from the teaching of the Chess Masters.

**Comment by kappa-wingman** [comment]
Good to hear that, thanks.

**Comment by SGQuek** [comment]
@kappa-wingman 
Thanks for verification. FYI, what you have tested is actually NewGG20180504 [ie: 4-May-2018] 
which utilizes CPU, instead of GPU like GGzero.

For your information, there has been a newer (third-party?) update of NewGG which is called 
NewGG20180519, distributed among many Chinese website for the last 2 years.
Because I can read Chinese, I managed to download it last year through a website.

It utilizes 128 threads and it is said to be even stronger than that NewGG20180504.

I would like to share this updated version in GitHub should it be legal to do so.

**Reply by kappa-wingman** [reply]
Yes, it's a standalone version of GGzero. It is the last public binary that could be get directly from Github. That version is already very strong and beats all other standalone versions of AI engines that I had tested.

That's cool that it also have a more advanced AI version (a distributed model).
Looking forward to see the legal releases in the future.

**Reply by mtaktikos** [reply]
As long as this engine is GPL licensed open source, there is no problem to share a compile of this. I have the exe compile of NewGG20180504 and would be interested to compare it with NewGG20180519

**Reply by SGQuek** [reply]
Here you go

The two zip files are actually the same.

Should there be any copyright issue, please let me know.
[ggX64set2018.zip](https://github.com/ianfab/Fairy-Stockfish/files/5780495/ggX64set2018.zip)
[NewGG20180519.zip](https://github.com/ianfab/Fairy-Stockfish/files/5780507/NewGG20180519.zip)

**Reply by mtaktikos** [reply]
Thank you

**Reply by mtaktikos** [reply]
In a first impression, only 2 games, NewGG20180519 looks in the middle game slightly better than NewGG20180504,  tournament file is attached. 
[NewGGvsNewGG.txt](https://github.com/ianfab/Fairy-Stockfish/files/5789892/NewGGvsNewGG.txt)

Both engines have nearly zero endgame knowledge. Their "sister" GGVip from the year 2012 seems to be the strongest in the endgame.
Unfortunately, it seems that nothing of the GG source codes can be used to tweak the Fairy-SF source code

**Reply by SGQuek** [reply]
Noted. I am recommending this updated version NewGG20180519 because it can support up to 128 thread rather than the 36 for NewGG20180504. So for those people who has a Ryzen threadripper (mine is 3970X), or the best Xeon processor, they may find that useful.
